src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs

```diff
@@ -97,74 +155,188 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
                         throw ExceptionUtilities.UnexpectedValue(tokenKind);
                 }
             }
-            sb.Append('$');
+            // If we're parsing a choice we should not exit without a closing '}'
+            return !parsingChoice;
+        }
 
-            return sb.ToString();
+        /// <summary>
+        /// Parses choice defined by the following grammar:
+        /// <![CDATA[
+        /// <choice> ::= "{" <choice-list> "}"
+        /// <choice-list> ::= <path-list> | <path-list> "," <choice-list>
+        /// ]]>
+        /// </summary>
+        private static bool TryCompileChoice(ref SectionNameLexer lexer, StringBuilder sb)
+        {
+            if (lexer.Lex() != TokenKind.OpenCurly)
+            {
+                return false;
+            }
+
+            // Start a non-capturing group for the choice
+            sb.Append("(?:");
+
+            // We start immediately after a '{'
+            // Try to compile the nested <path-list>
+            while (TryCompilePathList(ref lexer, sb, parsingChoice: true))
+            {
+                // If we've succesfully compiled a <path-list> the last token should
+                // have been a ',' or a '}'
+                char lastChar = lexer[lexer.Position - 1];
+                if (lastChar == ',')
+                {
+                    // Another option
+                    sb.Append("|");
+                }
+                else if (lastChar == '}')
+                {
+                    // Close out the capture group
+                    sb.Append(")");
+                    return true;
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(lastChar);
+                }
+            }
+
+            // Propagate failure
+            return false;
         }
 
-        private ref struct SectionNameLexer
+        /// <summary>
+        /// Parses range defined by the following grammar.
+        /// <![CDATA[
+        /// <range> ::= "{" <integer> ".." <integer> "}"
+        /// <integer> ::= "-" <digit-list> | <digit-list>
+        /// <digit-list> ::= <digit> | <digit> <digit-list>
+        /// <digit> ::= 0-9
+        /// ]]>
+        /// </summary>
+        private static (string numStart, string numEnd)? TryParseNumberRange(ref SectionNameLexer lexer)
+        {
+            var saved = lexer.Position;
+            if (lexer.Lex() != TokenKind.OpenCurly)
+            {
+                lexer.Position = saved;
+                return null;
+            }
+
+            var num1 = lexer.TryLexNumber();
+            if (num1 is null || lexer.IsDone)
+            {
+                // Not a number
+                lexer.Position = saved;
+                return null;
+            }
+
+            // The next two characters must be ".."
+            if (lexer.EatCurrentCharacter() != '.' || lexer.IsDone ||
```

Comment:
>EatCurrentCharacter() != '.' || lexer.IsDone [](start = 22, length = 44)

Consider creating a TryEatCurrentCharacter or similar so you don't have to keep checking lexer.IsDone everywhere. Seems fragile the way that it's done now.